pkg/controller/core/workload_controller.go

```diff
@@ -144,11 +154,30 @@ func (r *WorkloadReconciler) Reconcile(ctx context.Context, req ctrl.Request) (c
 	log.V(2).Info("Reconciling Workload")
 
 	if len(wl.ObjectMeta.OwnerReferences) == 0 && !wl.DeletionTimestamp.IsZero() {
+		// manual deletion triggered by the user
 		return ctrl.Result{}, workload.RemoveFinalizer(ctx, r.client, &wl)
```

Comment:
This will remove the workload finalizer when there is no Job owner. What removes the finalizer if the Job object still exists?

Reply:
I might be lucking some understanding here so please excuse my ignorance 🙏 

How I understand the problem is that the primary function of this code branch (which was not created by me, I just added a comment here to be able to immediately differentiate it from the rest of the deletion logic) is to act when a job was manually deleted by a user and we need to cleanup workload resources because kueue controls them and they won't be deleted by K8s itself and so that we could also control the cleanup of the kueue internal state [cache, internal queues). That's why there are 2 conditions here no owner (parent Job) and there's a deletion timestamp set so the workload itself is supposed to be deleted too.

Today if no one touches the Job, its Workload will stay. The primary function of the code in this PR is to cleanup K8s and kueue's knowledge about completed workloads to free up memory and storage, so the Workload needed to finish first and only then it could be deleted. My assumption is that when a Workload is completed it shouldn't have any finalizers (and resources) in place so that's why I didn't do anything with the finalizers. 

@mimowo If that is not the case and a completed Workload can indeed still have finalizers I think I'd need to some guidance with how to achieve this state and how it should be handled as I was not able to see it with my own basic tests.

Reply:
> My assumption is that when a Workload is completed it shouldn't have any finalizers (and resources) in place so that's why I didn't do anything with the finalizers.

Yes, this is correct, we should make sure the finalizer for a finished workload is deleted. I think this is the code: which already does it: https://github.com/kubernetes-sigs/kueue/blob/a128ae3354a56670ffa58695fef1eca270fe3a47/pkg/controller/jobframework/reconciler.go#L332-L340.  In that case calling `Delete` as you do is enough.

However, given that the code change depends on the logic in another part of the codebase I would prefer to have an integration test to show the workloads get deleted from the API server.